### PR TITLE
Add replication source field

### PR DESCRIPTION
**Pull request by ELENAGER**
We are adding support for replication on Volume Group level. In order to do so, we need to be able to receive either volume id or volume group id, and not only volume id, as it was done previously.
We are not changing an existing field volume_id for backward compatibility, but adding a new optional filed replication_source.

Signed-off-by: Elena Gershkovich <elenage@il.ibm.com>

### CODE REVIEW

File: replication/README.md [hunk 1]
```diff
@@ -104,6 +104,9 @@ message EnableVolumeReplicationRequest {
   // This field SHALL be used by the CO in subsequent calls to refer to
   // this volume.
   string volume_id = 1;
+  // If specified, this field will contain volume or volume group id
+  // for replication. This field is OPTIONAL.
+  ReplicationSource replication_source = 5;
```
Comment (Madhu-1) [inline replication/README.md:109]: Please arrange this in order. This should not be between 1 and 2. 

same is applicable for other places as well

**Review by Madhu-1 — COMMENTED** [review]
We had volume_id and replication_id in all the requests, and now we have replication_source. IMO we should not keep all these, and we need to deprecate the volume_id and use the replication_source as the main object to identify it's a volume replication or the group replication.

File: replication/replication.proto [hunk 2]
```diff
@@ -196,6 +211,9 @@ message GetVolumeReplicationInfoRequest {
   // This field SHALL be used by the CO in subsequent calls to refer to
   // this volume.
   string volume_id = 1;
+  // If specified, this field will contain volume or volume group id
+  // for replication. This field is OPTIONAL.
```
Comment (Madhu-1) [inline replication/replication.proto:215]: This cannot be optional, right? This need to be set to `VolumeSource` or `VolumeGroupSource` isn't it

File: replication/replication.proto [hunk 3]
```diff
@@ -213,3 +231,27 @@ message GetVolumeReplicationInfoResponse {
...
   .google.protobuf.Timestamp last_sync_time = 1;
 }
+// Specifies what source the replication will be created from. One of the
+// type fields MUST be specified.
+message ReplicationSource {
+  // Message containing volume id
+  message VolumeSource {
+    // Contains identity information for the existing volume.
+    // This field is REQUIRED.
+    string replication_volume_id = 1;
```
Comment (Madhu-1) [inline replication/replication.proto:241]: Keep this as volume_id?

File: replication/replication.proto [hunk 4]
```diff
@@ -213,3 +231,27 @@ message GetVolumeReplicationInfoResponse {
...
+  }
+
+  // Message containing volume group id
+  message VolumeGroupSource {
+    // Contains identity information for the existing volume group.
+    // This field is REQUIRED.
+    string replication_volume_group_id = 1;
```
Comment (Madhu-1) [inline replication/replication.proto:248]: This should be renamed as volume_group_id? No need to append replication to it.

File: replication/replication.proto [hunk 5]
```diff
@@ -213,3 +231,27 @@ message GetVolumeReplicationInfoResponse {
...
+// type fields MUST be specified.
+message ReplicationSource {
+  // Message containing volume id
+  message VolumeSource {
+    // Contains identity information for the existing volume.
+    // This field is REQUIRED.
+    string replication_volume_id = 1;
+  }
+
+  // Message containing volume group id
```
Comment (Madhu-1) [inline replication/replication.proto:244]: Remove this comment, as this is not the correct one.

File: replication/replication.proto [hunk 6]
```diff
@@ -213,3 +231,27 @@ message GetVolumeReplicationInfoResponse {
   // This field is REQUIRED.
   .google.protobuf.Timestamp last_sync_time = 1;
 }
+// Specifies what source the replication will be created from. One of the
+// type fields MUST be specified.
+message ReplicationSource {
+  // Message containing volume id
```
Comment (Madhu-1) [inline replication/replication.proto:237]: remove this comment

File: replication/replication.proto (hunk 2 above)
Comment (ELENAGER) [reply replication/replication.proto:215]: replication_source field is optional if we are preserving volume_id field for backwards compatibility. If we decide to deprecate the volume_id and use the replication_source as the main object to identify it's a volume replication or the group replication - it must be changed to mandatory.

File: replication/README.md (hunk 1 above)
Comment (ELENAGER) [reply replication/README.md:109]: I agree, but I see that also   string replication_id = 4 [(alpha_field) = true] is situated between 1 and 2. Do you want me to fix both replication_source and replication_id  placement?

**Comment by ELENAGER** [comment]
@Madhu-1 , I need your help in understanding, how the deprecation of the volume_id must look like. Also, replication_id field must be kept for mirroring replication - we are using it for replication handle of the mirrored volume id.

File: replication/README.md (hunk 1 above)
Comment (ELENAGER) [reply replication/README.md]: All fields are arranged according to their numbers.

File: replication/replication.proto (hunk 3 above)
Comment (ELENAGER) [reply replication/replication.proto]: Renamed to volume_id

File: replication/replication.proto (hunk 4 above)
Comment (ELENAGER) [reply replication/replication.proto]: Renamed to volume_group_id

File: replication/replication.proto (hunk 5 above)
Comment (ELENAGER) [reply replication/replication.proto]: Comment removed

File: replication/replication.proto (hunk 6 above)
Comment (ELENAGER) [reply replication/replication.proto]: Comment removed

File: replication/replication.proto [hunk 7]
```diff
@@ -213,3 +231,25 @@ message GetVolumeReplicationInfoResponse {
...
+    // Contains identity information for the existing volume group.
+    // This field is REQUIRED.
+    string volume_group_id = 1;
+  }
+
+  oneof type {
+    // Volume group source type
+    VolumeGroupSource volumegroup = 1;
+    // Volume source type
+    VolumeSource volume = 2;
```
Comment (Madhu-1) [inline replication/replication.proto:253]: keep the volume first and than group
```suggestion
   // Volume source type
   VolumeSource volume = 1;
   // Volume group source type
   VolumeGroupSource volumegroup = 2;
 
```

**Review by Madhu-1 — COMMENTED** [review]
changes LGTM, @ShyamsundarR @nixpanic any comments?

**Comment by Madhu-1** [comment]
> @Madhu-1 , I need your help in understanding, how the deprecation of the volume_id must look like. Also, replication_id field must be kept for mirroring replication - we are using it for replication handle of the mirrored volume id.

Sorry missed this one, @ELENAGER sure we can discuss on that one lets get this one merged. our idea was to deprecate both volume_id and replication_id and use below structure as the replacement for the replication_id (rough idea)

```go
// Specifies what destination the replication will be created for. One of the
message ReplicationDestination {
  message VolumeDestination {
    // Contains identity information for the volume on the remote cluster.
    string volume_id = 1;
  }

  message VolumeGroupDestination {
    // Contains identity information for the existing volume group on destination.
    string volume_group_id = 1;
  }

  oneof type {
    // Volume group destination type
    VolumeGroupDestination volumegroup = 1;
    // Volume destination type
    VolumeDestination volume = 2;
  }
}
```

**Review by Madhu-1 — COMMENTED** [review]
Please make the required minor change and squash the commits into one as we dont manually squash and merge the PR (its handled by a bot).

File: replication/replication.proto (hunk 7 above)
Comment (ELENAGER) [reply replication/replication.proto]: Done

**Pull request closed without merging**